cpp/include/cudf/io/experimental/hybrid_scan.hpp

```diff
@@ -0,0 +1,253 @@
+  /**
+   * @brief Constructor for the experimental parquet reader class to optimally read Parquet files
+   * subject to highly selective filters
+   *
+   * @param footer_bytes Host span of parquet file footer bytes
+   * @param options Parquet reader options
+   */
+  explicit hybrid_scan_reader(cudf::host_span<uint8_t const> footer_bytes,
+                              parquet_reader_options const& options);
+
+  /**
+   * @brief Destructor for the experimental parquet reader class
+   */
+  ~hybrid_scan_reader();
+
+  /**
+   * @brief Get the Parquet file footer metadata
+   *
+   * Returns the materialized Parquet file footer metadata struct. The footer will contain the
+   * materialized page index if called after `setup_page_index()`.
+   *
+   * @return Parquet file footer metadata
+   */
+  [[nodiscard]] FileMetaData const& get_parquet_metadata() const;
+
+  /**
+   * @brief Get the byte range of the page index in the Parquet file
+   *
+   * @return Byte range of the page index
+   */
+  [[nodiscard]] byte_range_info get_page_index_bytes() const;
+
+  /**
+   * @brief Setup the page index within the Parquet file metadata (`FileMetaData`) for later use
+   *
+   * Materialize the `ColumnIndex` and `OffsetIndex` structs (collectively called the page index)
+   * within the Parquet file metadata struct. The statistics contained in page index can be used to
+   * prune data pages before decoding
+   *
+   * @param page_index_bytes Host span of Parquet page index buffer bytes
+   */
+  void setup_page_index(cudf::host_span<uint8_t const> page_index_bytes) const;
+
+  /**
+   * @brief Get all available row groups from the parquet file
+   *
+   * @param options Parquet reader options
+   * @return Vector of row group indices
+   */
+  [[nodiscard]] std::vector<size_type> get_all_row_groups(
+    parquet_reader_options const& options) const;
+
+  /**
+   * @brief Filter the input row groups using column chunk statistics
+   *
+   * @param row_group_indices Input row groups indices
+   * @param options Parquet reader options
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @return Filtered row group indices
+   */
+  [[nodiscard]] std::vector<cudf::size_type> filter_row_groups_with_stats(
+    cudf::host_span<size_type const> row_group_indices,
+    parquet_reader_options const& options,
+    rmm::cuda_stream_view stream) const;
+
+  /**
+   * @brief Get byte ranges of bloom filters and dictionary pages (secondary filters) for
+   * further row group pruning
+   *
+   * @param row_group_indices Input row groups indices
+   * @param options Parquet reader options
+   * @return Pair of vectors of byte ranges to per-column-chunk bloom filters and dictionary pages
+   */
+  [[nodiscard]] std::pair<std::vector<byte_range_info>, std::vector<byte_range_info>>
+  get_secondary_filters(cudf::host_span<size_type const> row_group_indices,
+                        parquet_reader_options const& options) const;
```

Comment:
should we preemptively add a stream parameter?

Reply:
This is deliberate (per design doc) as this API is all host code so not taking in stream means GPU will be available to Spark during this time.